README

This is the README file for Planner.

Synopsis
========

Planner is an organizer and day planner for Emacs.  It helps you keep
track of your pending and completed tasks, daily schedule, dates to
remember, notes and inspirations.  It is a powerful tool not only for
managing your time and productivity, but also for keeping within easy
keystroke reach all of the information you need to be productive.  It
can even publish reports charting your work for your personal web
page, your conscience, or your soon-to-be-impressed boss.

In fact, because it uses as its building blocks simple plain-text
files, it is an incredibly modular and flexible tool capable of
shaping and handling your personal information in ways whose variety
is limited only by your imagination.  Because of this, Planner has a
very active and generous community who regularly share their
innovations with each other.  Many of these modules and extensions are
included.

License
=======

Planner is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2, or (at your option) any
later version.

Planner is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with Planner; see the file COPYING.  If not, write to the Free
Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301, USA.

Directory contents
==================

The directory structure is as follows.

top level :: Planner source code and the documentation
             (planner-el.texi).

contrib :: Files that are not a part of Planner proper, but are useful
           enough to be bundled with Planner.

scripts :: Scripts that are used when compiling Planner.

Getting started
===============

Compilation
-----------

This is an optional step, since Emacs Lisp source code does not
necessarily have to be byte-compiled.  It will yield a speed increase,
though.

A working copy of Emacs or XEmacs is needed in order to compile
Planner.  By default, the program that is installed with the name
`emacs' will be used.

If you want to use the `xemacs' binary to perform the compilation, you
would need to edit `Makefile.defs' in the top-level directory as
follows.  You can put either a full path to an Emacs or XEmacs binary
or just the command name, as long as it is in the PATH.

EMACS    = xemacs
SITEFLAG = -no-site-file

In order for Planner to be byte-compiled properly, you will need to
specify the paths to Muse and other external software programs in
`scripts/planner-build.el'.  If a file fails to be byte-compiled, only
its source will be installed.

Running `make' should compile the Planner source files in the `lisp'
directory.

Installation
------------

Planner may be installed into your file hierarchy by doing the
following.

Edit the `Makefile.defs' file so that ELISPDIR points to where you
want the source and compiled Planner files to be installed and INFODIR
indicates where to put the Planner manual.  Of course, you will want
to edit EMACS and SITEFLAG as shown in the Compilation section if you
are using XEmacs.

If you are installing Planner on a Debian system, you might want to
change the value of INSTALLINFO as specified in `Makefile.defs'.

If you wish to install Planner to different locations than the
defaults specify, edit `Makefile.defs' accordingly.

Run `make' as a normal user.

Run `make install' as the root user if you have chosen installation
locations that require this.

Insinuation
-----------

Two things need to happen in order for Planner to be usable with your
version of Emacs or XEmacs.

 1. The location of the Planner directory needs to be added to the load
    path so that your variant of Emacs can find it.

 2. You need to load whichever Planner files you wish to make use of.

A quick example that accomplishes both of these follows.

;; Add this to your .emacs or .xemacs/init.el file.
(add-to-list 'load-path "/path/to/planner")
(require 'planner)

Documentation
-------------

The Planner manual may be generated by running `make doc'.

It will produce two files: an Info file called `planner-el.info' and
an HTML document called `planner-el.html'.

This manual is also available online in several forms.

 - PDF: http://www.mwolson.org/static/doc/planner.pdf
 - HTML (single file): http://www.mwolson.org/static/doc/planner.html
 - HTML (multiple files): http://www.mwolson.org/static/doc/planner/

Further Documentation
=====================

Planner has an official website at http://www.plannerlove.com/.  It is
a collaborative wiki.

Bugs may be reported using the Planner Bug-Tracker at
https://gna.org/bugs/?group=planner-el.

A mailing list for Planner and some related software exists; consult
http://www.emacswiki.org/cgi-bin/wiki/EmacsWikiMailingList for details
on joining it or viewing its contents.


NEWS
Planner NEWS --- History of user-visible changes    -*- outline -*-

* Changes in Planner 3.41 (unreleased)

** Bugfixes

*** Make planner-visit-link more robust.
*** Don't insert too many spaces around task number.
*** Fix lock up with old delegated task mark.
*** planner-el.texi: Many updates and clarifications were added.
*** planner-appt: Fix highlighting in plan pages.
*** planner-authz: Work with latest changes in planner-publish.
*** planner-calendar: Fix many bugs and improve implementation.
Be sure to read the "Publishing Calendars" section of the manual for
details, as the common setup for Planner 3.40 will not work.
This feature should now be much easier to get working.
*** planner-calendar: If the current page is not a daypage,
default to publishing a calendar for today.
*** planner-gnus: Fix buggy browsing behavior.
*** planner-publish: Fix erroneous output at end of file.

** Compatibility fixes

*** planner-appt: Fix initialization for XEmacs.
*** planner-appt: For XEmacs, create a diary file if one does not exist.
*** planner-erc: Work with newer versions of ERC.
*** planner-psvn: Make this work with newer versions of psvn.el.

** New features

*** Display a helpful message if muse-project-alist is undefined.
*** New function `planner-tomorrow' returns the date for tomorrow.
*** planner-appt: New option `planner-appt-forthcoming-show-day-names-flag'
that indicates whether to show day names in forthcoming appointments.
*** planner-authz: Add planner-multi support.
*** planner-authz: Consider all of a note's links when building an
access list.
*** planner-authz: Only publish dependencies if they haven't already
been published.
*** planner-authz: Improve list handling.
*** planner-calendar: Make a <calendar> tag.
*** planner-calendar: Surround generated calendar with <div> tag.
*** planner-erc: Connect to an existing chat session when following a
link, if possible.
*** planner-id: Remove task IDs when publishing.
*** planner-publish: Enable multiple links with "categories" attribute.
*** planner-publish: Use <h2>, <h3>, etc. for headings.
*** planner-publish: Add default CSS elements for calendar, notes,
and diary sections.
*** planner-publish: Preserve line breaks when publishing diary section.
*** planner-report: Sort pages by plan page name.

** New files

*** planner-registry.el: Registry of URLs and Planner annotations.
*** planner-zoom.el: Navigate plan pages.

* Changes in Planner 3.40

** Migration to Muse backend from emacs-wiki is complete.
Planner is now based on Emacs Muse, rather than emacs-wiki.  This will
require a few changes to your Planner configuration.  See the
`Creating Your Planner' section of the Planner manual for an example
Planner and Muse configuration.

Other considerations follow.

*** WikiNames are not colorized by default.
If you want WikiName support, add a "(require 'muse-wiki)" line to
your Planner configuration.

*** List items that that begin with "- " should begin with " - ".

*** The use of `planner-option-customized' is deprecated.
Please change any use of `planner-option-customized' to `setq' in your
configuration.  Most of the options that required this no longer
exist.

*** Planner files must have a .muse extension.
Make sure that you rename your Planner files to have a ".muse"
extension, or they will not be recognized.

*** Publishing may be done by means of the planner-publish.el file.
Add "(require 'planner-publish)" to your Planner configuration to use
it.  This will make available several publishing styles that you can
use with `muse-project-alist'.

The difference between planner-specific publishing styles and the
normal Muse publishing styles is that the former can mark up tasks and
notes specially.

** New files

*** AUTHORS: List of contributors to Planner and authors of Planner files.
*** ChangeLog.2004, ChangeLog.2005, ChangeLog:
A fairly complete history of changes to Planner since January 2004.
*** COMMENTARY: John Wiegley's original Planner commentary,
moved from PLANNER-README.
*** NEWS: This file, which lists user-visible changes since Planner 3.28.
*** Makefile, Makefile.defs, scripts/planner-build.el:
A simple build system for Planner.
*** README: Instructions for building and installing Planner.

** Compatibility

*** Browsing Gnus URLs should work with Emacs21.
*** Several XEmacs fixes were committed.

** Misc

*** Change of Delegated mark from ">" to "D".
Make sure that you use "D" instead of ">" to indicate delegated tasks,
or these tasks will not display or publish properly.

*** planner-accomplishments: Line up accomplishments neatly.
*** planner-appt: Forthcoming appointments are now dealt with.
*** planner-authz: Several TODO items were completed.
*** planner-deadline: New `add' and `remove' functions.
*** planner-multi: New option `planner-multi-date-links-last-p'
indicates whether to put date links after links to other pages.
*** planner-rss: Some options might not be backwards-compatible.
*** planner-timeclock-summary-proj: Several enhancements were made.

*** planner.el

**** Don't insert task padding if `planner-use-task-numbers' is nil.
**** Shortened dates like "2005.1.6" work.
**** When no day pages are desired, more things work than before.
**** planner-initial-page: New option that determines which page to
visit by default when not using day pages.
**** Permit links to buffer names.

* Changes in Planner 3.31

** New or noteworthy

*** dev/402: New: Provide planner-log-edit like functionality for xtla
*** dev/403: New: planner-delete-section[-text], add CREATE to
planner-narrow-to-section
*** dev/411: New: planner-mark-task-hook
*** dev/417,418: New: planner-multi-copy-tasks-to-page
*** dev/420: planner-annotation-as-kill: New interactive prefix
*** dev/421: New: planner-annotation-from-file-with-position
*** dev/430: planner-update-task: Remove need for advice,
planner-multi-replan-task: Accept list as argument
*** dev/431: NOTE: planner-create-note: Default to current page

planner-mhe.el
;;; planner-mhe.el --- MH-E integration for the Emacs Planner

;; Copyright (C) 2004, 2005 Christophe Garion
;; Parts copyright (C) 2004, 2005 Free Software Foundation, Inc.

;; Author: Christophe Garion <garion@supaero.fr>
;; Author: Sandra Jean Chua <sacha@free.net.ph>
;; Created: <2004-08-09 17:16:57 tof planner-mhe.el>
;; Time-stamp: <12/04/2005 19:26:58 Yann Hodique>
;; Keywords: planner, mh-e

;; This file is part of Planner.  It is not part of GNU Emacs.

;; Planner is free software; you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; Planner is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with Planner; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; Inspired by planner-gnus (thanks to Sacha Chua)

;; This file adds annotations for MH-E messages. You will then be able
;; to use M-x planner-create-task-from-buffer to create tasks from
;; MH-E folder or show buffers with the correct annotation. If you
;; create an annotation from a mh-index folder, the message "real"
;; folder will be taken into account.

;;; Contributors:

;; Yann Hodique helped port this to Muse.

;;; Code:

(require 'planner)
(require 'mh-e)
(require 'mh-index)

(defun planner-mhe-get-message-folder-from-index ()
  "Returns the name of the message folder in a index folder
  buffer."
  (save-excursion
    (mh-index-previous-folder)
    (buffer-substring
     (planner-line-beginning-position)
     (planner-line-end-position))))

(defun planner-mhe-get-message-real-folder ()
  "Return the name of the current message real folder, so if you use
  sequences, it will now work."
  (save-excursion
    (let* ((folder
            (if (equal major-mode 'mh-folder-mode)
                mh-current-folder
              ;; Refer to the show buffer
              mh-show-folder-buffer))
           (end-index (min (length mh-index-folder) (length folder))))
      ;; a simple test on mh-index-data does not work, because
      ;; mh-index-data is always nil in a show buffer.
      (if (string= mh-index-folder (substring folder 0 end-index))
          (if (equal major-mode 'mh-show-mode)
              (save-window-excursion
                (when (buffer-live-p (get-buffer folder))
                  (progn
                    (pop-to-buffer folder)
                    (planner-mhe-get-message-folder-from-index))))
            (planner-mhe-get-message-folder-from-index))
        folder))))

(defun planner-mhe-get-message-folder ()
  "Return the name of the current message folder."
  (if (equal major-mode 'mh-folder-mode)
      mh-current-folder
    ;; Refer to the show buffer
    mh-show-folder-buffer))

(defun planner-mhe-get-message-field (field)
  "Return a particular field of the current message."
  (save-excursion
    (let ((num
           (if (equal major-mode 'mh-folder-mode)
               (mh-get-msg-num nil)
             ;; Refer to the show buffer
             (mh-show-buffer-message-number)))
          (folder (planner-mhe-get-message-folder)))
      (car (split-string
            (with-temp-buffer
              (call-process (expand-file-name "anno" mh-progs)
                            nil t nil
                            folder
                            "-list" "-component" field
                            (number-to-string num))
              (buffer-string))
            "\n")))))

;;;###autoload
(defun planner-mhe-annotation ()
  "If called from a MH-E folder or message buffer, return an annotation.
Suitable for use in `planner-annotation-functions'."
  (when (or (equal major-mode 'mh-folder-mode)
            (equal major-mode 'mh-show-mode))
    (let ((from-header (planner-mhe-get-message-field "From"))
          (to-header (planner-mhe-get-message-field "To")))
      (planner-make-link
       (concat "mhe://" (planner-mhe-get-message-real-folder) "/"
               (planner-mhe-get-message-field "Message-Id"))
       (concat "E-Mail "
               (if (and planner-ignored-from-addresses
                        from-header
                        (string-match planner-ignored-from-addresses
                                      from-header))
                   ;; Mail from me, so use the To: instead
                   (concat "to " (planner-get-name-from-address
                                  to-header))
                 ;; Mail to me, so use the From:
                 (concat "from " (planner-get-name-from-address
                                  from-header))))))))

;;;###autoload
(defun planner-mhe-browse-url (url)
  "If this is a MH-E URL, jump to it."
  (when (string-match "\\`mhe://\\(.+\\)/\\([^>]+\\)" url)
    (let* ((folder (match-string 1 url))
           (num (match-string 2 url))
           (show-buf (concat "show-" folder)))
      (save-window-excursion
        (mh-visit-folder folder)
        (get-buffer-create show-buf)
        (mh-display-msg
         (string-to-number
          (if (= (aref num 0) ?<) ; message-id
              (car (split-string
                    (with-temp-buffer
                      (call-process
                       (expand-file-name "pick" mh-progs)
                       nil t nil
                       folder
                       "--message-id"
                       num)
                      (buffer-string))
                    "\n"))
            num))
         folder))
      (pop-to-buffer show-buf))))

;(fset 'planner-get-from 'planner-gnus-get-address)
;(fset 'planner-get-message-id 'planner-gnus-get-message-id)
(custom-add-option 'planner-annotation-functions
                   'planner-mhe-annotation)
(add-hook 'planner-annotation-functions 'planner-mhe-annotation)
(planner-add-protocol "mhe://" 'planner-mhe-browse-url nil)

; to set the mh-path etc. variables
(mh-find-path)

(provide 'planner-mhe)
;;; planner-mhe.el ends here
